emscripten.py

```diff
@@ -127,6 +127,8 @@ def update_settings_glue(metadata, DEBUG):
 
   # start with the MVP features, and add any detected features.
   settings.BINARYEN_FEATURES = ['--mvp-features'] + metadata['features']
+  if settings.MEMORY64 and '--enable-memory64' not in settings.BINARYEN_FEATURES:
+    settings.BINARYEN_FEATURES += ['--enable-memory64']
```

Comment:
Why is this different than the other features, in that it checks if it already exists instead of asserting it exists? (comment would be good, as it seems surprising when reading)

Reply:
This is not like an assert? More like making sure it is not added twice..
I'll add a comment.

Reply:
Is this needed though?   Shouldn't `--enable-memory64` always be enabled if any of the input object files are built with `MEMORY64`?    Can we instead assert like below?

Reply:
Would that have to be set by the user? Because currently it is not set anywhere else. I was assuming that it should be possible for someone to compile with just `-s MEMORY64=2` and not have to fiddle with Binaryen settings manually.

Reply:
Sorry if I wasn't clear enough before. This line looks like
```python
if X and not flag in features:
  features += flag
```
while the lines right under it all look like this:
```python
if X:
  assert flag in features
```
It seems surprising to have them handled differently, so a comment explaining that could help I think. (As to the reason, perhaps we set USING_PTHREADS in emcc.py based on `-pthread`, which would explain why we can assume it is set already?)

Also, I see that in fact two ifs down is an if for MEMORY64, which is redundant with this one?

At least the redundancy seems important to me to fix. If the other issue is unclear, a TODO to harmonize things would be enough for me.

Reply:
I think they idea is that if any object files were built with `MEMORY64` then `BINARYEN_FEATURES` should already contains `--enable-memory64` since `BINARYEN_FEATURES` I derived from running binaryen with `--detect-features`.

Reply:
Ok, I will make it just assert then. This seemed to be necessary in the past, but I guess we'll find out later if that is still the case.

Reply:
Hmm maybe the problem is that `--detect-features` will think its a wasm32 module after `--memory64-lowering` ?

Reply:
That could be true yes.. `--detect-features` is used always on the very first call to a binaryen tool which is always `wasm-emscripten-finalize` I think.   Is that also where you do the lowering?  

@tlively WDYT, should `--detect-features` report features that are lowered away?

Reply:
That depends what it is used for.. a lowered module should be treated as wasm32 by any binary tools, but as wasm64 for the JS bindings.

Reply:
Hmm.. actually if this is after `--memory64-lowering` has happened then presumably we don't need to (or want to) run with `--enable-memory64`... once its been lowered away we don't want binaryen re-introducing it.. or considering it legal.

Reply:
So.. that means we can't have the assert either, since it will fail, because `settings.MEMORY64` was still true.
Either way, that will involve further testing, for the purpose of getting those other PRs landed, can we leave that till later?

Reply:
I think you can just remove these two lines.. I don't think they are needed or useful.

Reply:
Yeah I think it would be good to have the lowering in Binaryen completely remove the feature so that `--print-features` no longer reports it as enabled. That way later tool invocations will not accidentally reintroduce it, as @sbc100 said.